practice-two/src/components/FormInput/FormInput.jsx

```diff
@@ -0,0 +1,60 @@
+import { useState } from 'react';
+import './FormInput.css';
+import PropTypes from 'prop-types';
+import { MESSAGES, REGEX } from '@constants';
+
+const FormInput = (props) => {
+  const [focused, setFocused] = useState(false);
+  const { label, errorMessage, onChange, ...inputProps } = props;
+
+  const handleFocus = () => {
```

Comment:
do you read the optimate performance I've said with you yesterday?

Reply:
Yes, I read it
- **memo**: 
  - It's a HOC, when parent component re-render, it will cause children components to be re-rendered.
  - `memo` will use for children components to avoid re-rendering when parent component re-render or prop unchanged.
  
- **useMemo**: 
  - It's used for avoiding some functions run again with no meaning when component re-render.
  - When component render first time, the result will store in cache.
  - When it re-render, useMemo will compare the result was stored in cache, if result is difference with result in cache, the function will do its job. Otherwise, it's the opposite.

- **useCallback**:
  - Even you use `memo` to avoiding re-render children components, but it's not working.
  - Should use `useCallback` when your children component uses `memo`.
  - `useCallback` usually uses for function

Reply:
just read but not apply?